### PR TITLE
#379 Deployment overhaul

### PR DESCRIPTION
**Pull request by julienrf**
This is an attempt to make the deployment simpler. The main changes are:
- the app are built by the CI (using the `sbt data/packageBin` command, for instance), rather than by the `run.sh` script ;
- there is a `Deploy.sc` script which unpacks the tarballs at the right place on the server.

### CODE REVIEW

File: data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala [hunk 1]
```diff
@@ -32,13 +30,4 @@ object BintrayMeta extends BintrayProtocol {
       .sortBy(_.created)(Descending)
   }
 
-  def append(paths: DataPaths, meta: BintraySearch): Unit = {
```
Comment (MasseGuillaume) [inline data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala]: hum this should be used somewhere.

File: data/src/main/scala/ch.epfl.scala.index.data/Main.scala [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import ch.epfl.scala.index.data.github.GithubDownload
+import ch.epfl.scala.index.data.maven.DownloadParentPoms
+import org.slf4j.LoggerFactory
+
+import scala.sys.process.Process
+
+/**
+  * This application manages indexed POMs.
+  */
+object Main {
```
Comment (julienrf) [inline data/src/main/scala/ch.epfl.scala.index.data/Main.scala]: I basically moved the `BintrayPipeline` into a `Main`, to make it clearer that it’s the app entry point.

File: data/src/main/scala/ch.epfl.scala.index.data/Main.scala [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            .fold(sys.error(s"Unknown step: $name. Available steps are: ${steps.map(_.name).mkString(" ")}."))(List(_))
+        case None        => sys.error(s"No step to execute. Available steps are: ${steps.map(_.name).mkString(" ")}.")
+      }
+
+    inPath(paths.contrib) { sh =>
+      logger.info("Pulling the latest data from the 'contrib' repository")
+      sh.exec("git", "checkout", "master")
+      sh.exec("git", "remote", "update")
+      sh.exec("git", "pull", "origin", "master")
+    }
```
Comment (julienrf) [inline data/src/main/scala/ch.epfl.scala.index.data/Main.scala]: Git repositories manipulation now happens here rather than in the `Run.sc` script.

File: data/src/main/scala/ch.epfl.scala.index.data/Main.scala [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    logger.info("Executing steps")
+    stepsToRun.foreach(_.run())
+
+    inPath(paths.index) { sh =>
+      logger.info("Saving the updated state to the 'index' repository")
+      sh.exec("git", "add", "-A")
+      def now() = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss"))
+      sh.exec("git", "commit", "--allow-empty", "-m", '"' +: now() :+ '"')
+    }
```
Comment (julienrf) [inline data/src/main/scala/ch.epfl.scala.index.data/Main.scala]: Note that we don’t push our changes. That’s only to make development easier: we want only the actual server to push its `ìndex` content to the github scaladex-index repository.

File: data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala (hunk 1 above)
Comment (julienrf) [inline data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala]: Where?

File: script/Deploy.sc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+def runSlurp(args: String*): String =
+  Process(args.toList).lineStream.toList.headOption.getOrElse("")
+
+/**
+  * @param dataTarball   Path of the packaged (.zip) `data` application
+  * @param serverTarball Path of the packaged (.zip) `server` application
+  * @param reposDir      Directory that contains the `scaladex-credentials`, `scaladex-index` and
+  *                      `scaladex-contrib` sub-directories containing the git repositories
+  */
+@main def main(dataTarball: String, serverTarball: String, reposDir: String) = {
```
Comment (julienrf) [inline script/Deploy.sc]: This is the deployment script to run on the server. We would eventually move it to an sbt task, but I leave it to another PR.

File: script/Run.sc [hunk 6]
```diff
@@ -1,121 +1,31 @@
...
 
 /**
-  * @param reposDir Directory that contains the `scaladex-credentials`, `scaladex-index` and `scaladex-contrib`
-  *                 sub-directories containing the git repositories
+  * @param dataPath Path of the deployed `data` application
+  * @param reposDir Directory that contains the `scaladex-credentials`, `scaladex-index` and
+  *                 `scaladex-contrib` sub-directories containing the git repositories
   */
-@main def main(reposDir: String, fullBranchName: String, job: Job.Value) = {
+@main def main(dataPath: String, reposDir: String, job: Job.Value) = {
```
Comment (julienrf) [inline script/Run.sc]: This script is useful to trigger a re-indexing or to re-create the ElasticSearch database.

File: data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala (hunk 1 above)
Comment (julienrf) [inline data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala]: It seems that it’s been superseeded by `writeMergedPoms` [here](https://github.com/scalacenter/scaladex/blob/master/data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayListPoms.scala#L132-L141).

Comment (MasseGuillaume) [inline data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala]: https://github.com/scalacenter/scaladex/blob/master/data/src/main/scala/ch.epfl.scala.index.data/Meta.scala#L65

Comment (julienrf) [inline data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala]: What’s your point?

Comment (MasseGuillaume) [inline data/src/main/scala/ch.epfl.scala.index.data/bintray/BintrayMeta.scala]: This is where it was refactored. I forgot to remove the above `append` method in the refactor.

**Comment by MasseGuillaume** [comment]
Thanks !
